compiler/rustc_middle/src/ty/inhabitedness/mod.rs

```diff
@@ -1,230 +1,101 @@
-    fn uninhabited_from(
-        &self,
-        tcx: TyCtxt<'tcx>,
-        substs: SubstsRef<'tcx>,
-        is_enum: bool,
-        param_env: ty::ParamEnv<'tcx>,
-    ) -> DefIdForest<'tcx> {
-        let data_uninhabitedness = move || self.ty(tcx, substs).uninhabited_from(tcx, param_env);
-        if is_enum {
-            data_uninhabitedness()
-        } else {
-            match self.vis {
-                Visibility::Restricted(from) => {
-                    let forest = DefIdForest::from_id(from);
-                    let iter = Some(forest).into_iter().chain(Some(data_uninhabitedness()));
-                    DefIdForest::intersection(tcx, iter)
+/// N.B. this query should only be called through `Ty::uninhabited_predicate`
+fn uninhabited_predicate<'tcx>(tcx: TyCtxt<'tcx>, ty: Ty<'tcx>) -> UninhabitedPredicate<'tcx> {
+    match *ty.kind() {
+        Adt(adt, substs) => {
+            if let Some(def_id) = adt.did().as_local() {
+                if matches!(tcx.representability(def_id), ty::Representability::Infinite) {
+                    return UninhabitedPredicate::False;
                 }
-                Visibility::Public => data_uninhabitedness(),
             }
+            UninhabitedPredicate::all(
+                tcx,
+                adt.variants()
+                    .iter()
+                    .map(|variant| variant.uninhabited_predicate(tcx, substs, adt)),
+            )
         }
-    }
-}
-
-impl<'tcx> Ty<'tcx> {
-    /// Calculates the forest of `DefId`s from which this type is visibly uninhabited.
-    fn uninhabited_from(
-        self,
-        tcx: TyCtxt<'tcx>,
-        param_env: ty::ParamEnv<'tcx>,
-    ) -> DefIdForest<'tcx> {
-        tcx.type_uninhabited_from(param_env.and(self))
-    }
-}
-
-// Query provider for `type_uninhabited_from`.
-pub(crate) fn type_uninhabited_from<'tcx>(
-    tcx: TyCtxt<'tcx>,
-    key: ty::ParamEnvAnd<'tcx, Ty<'tcx>>,
-) -> DefIdForest<'tcx> {
-    let ty = key.value;
-    let param_env = key.param_env;
-    match *ty.kind() {
-        Adt(def, substs) => def.uninhabited_from(tcx, substs, param_env),
-
-        Never => DefIdForest::full(),
 
         Tuple(ref tys) => {
-            DefIdForest::union(tcx, tys.iter().map(|ty| ty.uninhabited_from(tcx, param_env)))
+            UninhabitedPredicate::any(tcx, tys.iter().map(|ty| ty.uninhabited_predicate(tcx)))
         }
 
-        Array(ty, len) => match len.try_eval_usize(tcx, param_env) {
-            Some(0) | None => DefIdForest::empty(),
-            // If the array is definitely non-empty, it's uninhabited if
-            // the type of its elements is uninhabited.
-            Some(1..) => ty.uninhabited_from(tcx, param_env),
+        Array(ty, len) => match len.kind().try_to_machine_usize(tcx) {
+            None => ty.uninhabited_predicate(tcx).and(tcx, UninhabitedPredicate::ConstNonZero(len)),
+            Some(0) => UninhabitedPredicate::False,
```

Comment:
These two arms are just an optimization, right? We could always run the None arm and it would work. Leave a comment to that regard